modules/host/contractmanager/storagefolder.go

```diff
@@ -0,0 +1,370 @@
+
+	// errRelativePath is returned if a path must be absolute.
+	errRelativePath = errors.New("storage folder paths must be absolute")
+)
+
+// storageFolder contains the metadata for a storage folder, including where
+// sectors are being stored in the folder. What sectors are being stored is
+// managed by the contract manager's sectorLocations map.
+type storageFolder struct {
+	// Progress statistics that can be reported to the user. Typically for long
+	// running actions like adding or resizing a storage folder.
+	atomicProgressNumerator   uint64
+	atomicProgressDenominator uint64
+
+	// Disk statistics for this boot cycle.
+	atomicFailedReads      uint64
+	atomicFailedWrites     uint64
+	atomicSuccessfulReads  uint64
+	atomicSuccessfulWrites uint64
+
+	// The index, path, and usage are all saved directly to disk.
+	index uint16
+	path  string
+	usage []uint64
+
+	// availableSectors indicates sectors which are marked as consumed in the
+	// usage field but are actually available. They cannot be marked as free in
+	// the usage until the action which freed them has synced to disk, but the
+	// settings should mark them as free during syncing.
+	//
+	// sectors is a count of the number of sectors in use according to the
+	// usage field.
+	availableSectors map[sectorID]uint32
+	sectors          uint64
+
+	// mu needs to be RLocked to safetly write new sectors into the storage
+	// folder. mu needs to be Locked when the folder is being added, removed,
+	// or resized.
+	mu sync.TryRWMutex
+
+	// An open file handle is kept so that writes can easily be made to the
+	// storage folder without needing to grab a new file handle. This also
+	// makes it easy to do delayed-syncing.
+	metadataFile file
+	sectorFile   file
+}
+
+// mostSignificantBit returns the index of the most significant bit of an input
+// value.
+func mostSignificantBit(i uint64) uint64 {
+	if i == 0 {
+		panic("no bits set in input")
+	}
+
+	bval := []uint64{0, 0, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7}
+	r := uint64(0)
+	if i&0xffffffff00000000 != 0 {
+		r += 32
+		i = i >> 32
+	}
+	if i&0x00000000ffff0000 != 0 {
+		r += 16
+		i = i >> 16
+	}
+	if i&0x000000000000ff00 != 0 {
+		r += 8
+		i = i >> 8
+	}
+	if i&0x00000000000000f0 != 0 {
+		r += 4
+		i = i >> 4
+	}
+	return r + bval[i]
+}
+
+// randFreeSector will take a usage array and find a random free sector within
+// the usage array. The uint32 indicates the index of the sector within the
+// usage array.
+func randFreeSector(usage []uint64) (uint32, error) {
+	// Pick a random starting location. Scanning the sector in a short amount
+	// of time requires starting from a random place.
+	start, err := crypto.RandIntn(len(usage))
+	if err != nil {
+		panic(err)
+	}
+
+	// Find the first element of the array that is not completely full.
+	var i int
+	for i = start; i < len(usage); i++ {
+		if usage[i] != math.MaxUint64 {
+			break
+		}
+	}
```

Comment:
Randomness is only for performance. If we start at the beginning every time we will have to scan approximately all full sectors before finding an opening. But if we start in a random place, and gaps are approx. randomly distributed, we only have to scan (fill %) items. Unlike truly random choices, this has a firm upper bound on scan time.